File: packages/common/src/types/game/status.ts
import type {
  ManagerRecap,
  Player,
  PlayerRecap,
  QuestionMedia,
  QuestionType,
  RoundRecapAward,
  TeamStanding,
} from "@razzoozle/common/types/game"

export const STATUS = {
  // NOTE: the literal type is widened to include "PAUSED" (runtime value is
  // unchanged) so existing web records keyed off STATUS.* (e.g. MANAGER_SKIP_BTN,
  // indexed by a raw Status) keep typechecking before WP-WEB-QR-PAUSE adds a
  // real PAUSED branch. Additive/back-compat only.
  SHOW_ROOM: "SHOW_ROOM" as "SHOW_ROOM" | "PAUSED",
  SHOW_START: "SHOW_START",
  SHOW_PREPARED: "SHOW_PREPARED",
  SHOW_QUESTION: "SHOW_QUESTION",
  SELECT_ANSWER: "SELECT_ANSWER",
  SHOW_RESULT: "SHOW_RESULT",
  SHOW_RESPONSES: "SHOW_RESPONSES",
  // Manager-only interstitial: the per-round recap highlights get their OWN
  // full-screen page (reusing RecapSequence) BETWEEN the answer reveal and the
  // leaderboard. Players never receive it.
  SHOW_ROUND_RECAP: "SHOW_ROUND_RECAP",
  SHOW_LEADERBOARD: "SHOW_LEADERBOARD",
  FINISHED: "FINISHED",
  WAIT: "WAIT",
  PAUSED: "PAUSED",
} as const

export type Status = (typeof STATUS)[keyof typeof STATUS]

export interface CommonStatusDataMap {
  SHOW_START: { time: number; subject: string }
  SHOW_PREPARED: { totalAnswers: number; questionNumber: number }
  SHOW_QUESTION: {
    question: string
    // Answer TEXTS only (no solutions/correct — anti-cheat). Display-only for the
    // presenter big-screen tiles; absent for slider questions (no discrete answers).
    answers?: string[]
    // Optional permutation of answer indices when randomizeAnswers is enabled.
    // Absent for slider questions or when randomize is off. Lets the client render
    // tiles in displayOrder while keeping canonical indices for scoring.
    displayOrder?: number[]
    media?: QuestionMedia
    cooldown: number
    submittedBy?: string
  }
  SELECT_ANSWER: {
    question: string
    answers?: string[]
    media?: QuestionMedia
    time: number
    totalPlayer: number
    // Slider questions (no `correct` here — must not leak to players)
    type?: QuestionType
    min?: number
    max?: number
    step?: number
    unit?: string
    // Sentence-builder questions: shuffled word chips (no solution info).
    shuffledChunks?: string[]
    // Low-latency mode: server-authoritative timing anchors. All OPTIONAL —
    // absent in normal mode and ignored by old clients. Client uses these only
    // to drive the countdown/UI (never for scoring — scoring is server-side).
    serverSeq?: number
    serverNowMs?: number
    questionStartAtServerMs?: number
    answerDeadlineAtServerMs?: number
    submittedBy?: string
  }
  SHOW_RESULT: {
    correct: boolean
    message: string
    points: number
    myPoints: number
    rank: number
    aheadOfMe: string | null
    // Bonus / streak display
    streak?: number
    streakBonus?: boolean
    bonus?: boolean
    firstCorrect?: boolean
    poll?: boolean
    achievements?: string[]
    // Sum of per-achievement bonus points unlocked this round (already folded
    // into `myPoints`). Present only when > 0; absent/0 in the shipped default.
    bonusPoints?: number
    // Total players in this game, so the client can suppress a hollow "1st
    // place" label in a solo (single-player) game (W1-D FIX 2).
    playerCount?: number
    // The revealed correct answer for the round, shown on the wrong-answer
    // screen. Present only when applicable (never for poll/correct).
    correctAnswer?: string
    // Sentence-builder: authored correct order, revealed after the round.
    correctChunks?: string[]
    // Auto-mode: ms until the screen auto-advances, so the client can render a
    // local countdown. OPTIONAL — present only while auto-mode is on and an
    // advance is armed; absent (manual mode / old clients ignore it) otherwise.
    autoAdvanceMs?: number
    // Per-round recap awards (fastest finger, first correct, streak, …). OPTIONAL
    // + additive: up to 3 awards computed server-side per round; old clients
    // ignore it. Same value on every player's SHOW_RESULT (game-wide highlights).
    roundRecap?: RoundRecapAward[]
    // Scoring mode used in this game ('speed' for time-decay, 'accuracy' for
    // full base points on correct). OPTIONAL/additive; old clients ignore it.
    scoringMode?: "speed" | "accuracy"
  }
  WAIT: { text: string; teamMode?: boolean }
  PAUSED: { reason?: string }
  FINISHED: {
    subject: string
    top: Player[]
    rank?: number
    teamStandings?: TeamStanding[]
    // Post-game recap / awards (WP-A). OPTIONAL + filled DIFFERENTLY per
    // recipient: the manager emit carries a `ManagerRecap` (the full awards list
    // for the big screen), the per-player emit carries a `PlayerRecap` (this
    // player's own card + the single award they won). Old clients ignore it.
    recap?: ManagerRecap | PlayerRecap
    // Auto-mode flag echoed on the end-game screen so the client knows whether
    // the host auto-advanced through the game. OPTIONAL (old clients ignore it).
    autoMode?: boolean
  }
}

interface ManagerExtraStatus {
  SHOW_ROOM: { text: string; inviteCode?: string }
  SHOW_RESPONSES: {
    question: string
    responses: Record<number, number>
    solutions: number[]
    answers: string[]
    media?: QuestionMedia
    // Slider result
    type?: QuestionType
    correct?: number
    unit?: string
    averageGuess?: number
    // Type-answer result (manager-only — never sent to players). Normalized
    // text -> count, the authored accepted answers, and the comparison mode.
    textResponses?: Record<string, number>
    acceptedAnswers?: string[]
    matchMode?: "exact" | "normalized" | "fuzzy"
    // Sentence-builder result reveal (manager-only; same authored order as
    // player SHOW_RESULT.correctChunks after answers are closed).
    correctChunks?: string[]
    // Per-round recap awards (same as players see on SHOW_RESULT) so the
    // manager can display the round highlights during answer statistics.
    // OPTIONAL + additive.
    roundRecap?: RoundRecapAward[]
  }
  // Manager-only per-round recap screen (its OWN full-screen page). Carries the
  // same RoundRecapAward[] the players see inline on SHOW_RESULT.
  SHOW_ROUND_RECAP: { roundRecap: RoundRecapAward[] }
  SHOW_LEADERBOARD: {
    oldLeaderboard: Player[]
    leaderboard: Player[]
    teamStandings?: TeamStanding[]
    // Auto-mode: ms until the leaderboard auto-advances to the next question, so
    // the client can render a local countdown. OPTIONAL — present only while
    // auto-mode is on and an advance is armed; absent otherwise.
    autoAdvanceMs?: number
    // Per-round recap awards (same array the players get on SHOW_RESULT) so the
    // manager big-screen can show the round highlights. OPTIONAL + additive.
    roundRecap?: RoundRecapAward[]
  }
}

export type PlayerStatusDataMap = CommonStatusDataMap

export type ManagerStatusDataMap = CommonStatusDataMap & ManagerExtraStatus

export type StatusDataMap = PlayerStatusDataMap & ManagerStatusDataMap

File: rust/protocol/src/status.rs
//! status.rs — OWNS: Status enum + StatusDataMap discriminated union
//! (the game:status / GAME.STATUS payload), plus RoundRecapAward.
//!

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use ts_rs::TS;

// Import Player from sibling module
use crate::player::Player;
use crate::quizz::QuestionMedia;

// RoundRecapKey enum
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
pub enum RoundRecapKey {
    FastestFinger,
    FirstCorrect,
    Streak,
    HighestRoundScore,
    RankClimber,
    AchievementUnlock,
    SlowestPlayer,
    MostWrong,
}

/// RoundRecapAward for per-round highlights
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct RoundRecapAward {
    pub key: RoundRecapKey,
    pub winner_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub winner_avatar: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<i32>,
}

// Status enum
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, TS)]
pub enum Status {
    #[serde(rename = "SHOW_ROOM")]
    ShowRoom,
    #[serde(rename = "SHOW_START")]
    ShowStart,
    #[serde(rename = "SHOW_PREPARED")]
    ShowPrepared,
    #[serde(rename = "SHOW_QUESTION")]
    ShowQuestion,
    #[serde(rename = "SELECT_ANSWER")]
    SelectAnswer,
    #[serde(rename = "SHOW_RESULT")]
    ShowResult,
    #[serde(rename = "SHOW_RESPONSES")]
    ShowResponses,
    #[serde(rename = "SHOW_ROUND_RECAP")]
    ShowRoundRecap,
    #[serde(rename = "SHOW_LEADERBOARD")]
    ShowLeaderboard,
    #[serde(rename = "FINISHED")]
    Finished,
    #[serde(rename = "WAIT")]
    Wait,
    #[serde(rename = "PAUSED")]
    Paused,
}

// Data structs for each status variant

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct ShowRoomData {
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub invite_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team_mode: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct ShowStartData {
    pub time: i32,
    pub subject: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct ShowPreparedData {
    pub total_answers: i32,
    pub question_number: i32,
}


#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct ShowQuestionData {
    pub question: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub answers: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_order: Option<Vec<i32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub media: Option<QuestionMedia>,
    pub cooldown: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub submitted_by: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct SelectAnswerData {
    pub question: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub answers: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub media: Option<QuestionMedia>,
    pub time: i32,
    pub total_player: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "type")]
    pub question_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub step: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unit: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub shuffled_chunks: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub server_seq: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub server_now_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub question_start_at_server_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub answer_deadline_at_server_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub submitted_by: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct ShowResultData {
    pub correct: bool,
    pub message: String,
    pub points: i32,
    pub my_points: i32,
    pub rank: i32,
    pub ahead_of_me: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub streak: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub streak_bonus: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bonus: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub first_correct: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub poll: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub achievements: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bonus_points: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub player_count: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correct_answer: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correct_chunks: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_advance_ms: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub round_recap: Option<Vec<RoundRecapAward>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scoring_mode: Option<ScoringMode>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "lowercase")]
pub enum ScoringMode {
    Speed,
    Accuracy,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct ShowResponsesData {
    pub question: String,
    pub responses: HashMap<String, i32>,
    pub solutions: Vec<i32>,
    pub answers: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub media: Option<QuestionMedia>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "type")]
    pub question_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correct: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unit: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub average_guess: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text_responses: Option<HashMap<String, i32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accepted_answers: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub match_mode: Option<MatchMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correct_chunks: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub round_recap: Option<Vec<RoundRecapAward>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "lowercase")]
pub enum MatchMode {
    Exact,
    Normalized,
    Fuzzy,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct ShowRoundRecapData {
    pub round_recap: Vec<RoundRecapAward>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct ShowLeaderboardData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub old_leaderboard: Option<Vec<Player>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub leaderboard: Option<Vec<Player>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team_standings: Option<Vec<TeamStanding>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_advance_ms: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub round_recap: Option<Vec<RoundRecapAward>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct TeamStanding {
    pub team_id: String,
    pub points: i32,
    pub player_count: i32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FinishedData {
    pub subject: String,
    pub top: Vec<Player>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rank: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team_standings: Option<Vec<TeamStanding>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recap: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_mode: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct WaitData {
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team_mode: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
pub struct PausedData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

/// GameStatus discriminated union: status name + data
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "name", content = "data")]
pub enum GameStatus {
    #[serde(rename = "SHOW_ROOM")]
    ShowRoom(ShowRoomData),
    #[serde(rename = "SHOW_START")]
    ShowStart(ShowStartData),
    #[serde(rename = "SHOW_PREPARED")]
    ShowPrepared(ShowPreparedData),
    #[serde(rename = "SHOW_QUESTION")]
    ShowQuestion(ShowQuestionData),
    #[serde(rename = "SELECT_ANSWER")]
    SelectAnswer(SelectAnswerData),
    #[serde(rename = "SHOW_RESULT")]
    ShowResult(ShowResultData),
    #[serde(rename = "SHOW_RESPONSES")]
    ShowResponses(ShowResponsesData),
    #[serde(rename = "SHOW_ROUND_RECAP")]
    ShowRoundRecap(ShowRoundRecapData),
    #[serde(rename = "SHOW_LEADERBOARD")]
    ShowLeaderboard(ShowLeaderboardData),
    #[serde(rename = "FINISHED")]
    Finished(FinishedData),
    #[serde(rename = "WAIT")]
    Wait(WaitData),
    #[serde(rename = "PAUSED")]
    Paused(PausedData),
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_show_start_roundtrip() {
        let json = json!({"name": "SHOW_START", "data": {"time": 5000, "subject": "Quiz"}});
        let status: GameStatus = serde_json::from_value(json.clone()).unwrap();
        match &status {
            GameStatus::ShowStart(d) => assert_eq!(d.time, 5000),
            _ => panic!(),
        }
        assert_eq!(serde_json::to_value(&status).unwrap(), json);
    }

    #[test]
    fn test_show_result_with_recap() {
        let json = json!({
            "name": "SHOW_RESULT",
            "data": {
                "correct": true,
                "message": "OK",
                "points": 100,
                "myPoints": 500,
                "rank": 1,
                "aheadOfMe": null,
                "roundRecap": [{
                    "key": "fastest_finger",
                    "winnerName": "Alice",
                    "value": 1000
                }],
                "scoringMode": "speed"
            }
        });
        let status: GameStatus = serde_json::from_value(json.clone()).unwrap();
        match &status {
            GameStatus::ShowResult(d) => {
                assert!(d.correct);
                assert_eq!(d.round_recap.as_ref().unwrap().len(), 1);
            }
            _ => panic!(),
        }
        assert_eq!(serde_json::to_value(&status).unwrap(), json);
    }
}
